unused_deps/unused_deps.go

```diff
@@ -0,0 +1,295 @@
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		log.Printf("reading %s: %s", paramsFileName, err)
+	}
+	return depsByJar
+}
+
+// unusedDeps returns a set of labels that are unused deps.
+// It reads Dependencies proto messages from depsFileName (a jdeps file), which indicate deps used
+// at compile time, and returns those values in the depsByJar map that aren't used at compile time.
+func unusedDeps(depsFileName string, depsByJar map[string]string) (unusedDeps map[string]bool) {
+	unusedDeps = make(map[string]bool)
+	data, err := ioutil.ReadFile(depsFileName)
+	if err != nil {
+		log.Println(err)
+		return unusedDeps
+	}
+	dependencies := &depspb.Dependencies{}
+	if err := proto.Unmarshal(data, dependencies); err != nil {
+		log.Println(err)
+		return unusedDeps
+	}
+	for _, label := range depsByJar {
+		unusedDeps[label] = true
+	}
+	for _, dependency := range dependencies.Dependency {
+		if *dependency.Kind == depspb.Dependency_EXPLICIT {
+			delete(unusedDeps, depsByJar[*dependency.Path])
+		}
+	}
+	return unusedDeps
+}
+
+// parseBuildFile tries to read and parse the contents of buildFileName.
+func parseBuildFile(buildFileName string) (buildFile *build.File, err error) {
+	data, err := ioutil.ReadFile(buildFileName)
+	if err != nil {
+		return nil, err
+	}
+	return build.Parse(buildFileName, data)
+}
+
+// getDepsExpr tries to parse the content of buildFileName and return the deps Expr for ruleName.
+func getDepsExpr(buildFileName string, ruleName string) build.Expr {
+	buildFile, err := parseBuildFile(buildFileName)
+	if buildFile == nil {
+		log.Printf("%s when parsing %s", err, buildFileName)
+		return nil
+	}
+	rule := edit.FindRuleByName(buildFile, ruleName)
+	if rule == nil {
+		log.Printf("%s not found in %s", ruleName, buildFileName)
+		return nil
+	}
+	depsExpr := rule.Attr("deps")
+	if depsExpr == nil {
+		log.Printf("no deps attribute for %s in %s", ruleName, buildFileName)
+	}
+	return depsExpr
+}
+
+// hasRuntimeComment returns true if expr has an EOL comment containing the word "runtime".
+// TODO(bazel-team): delete when this comment convention is extinct
+func hasRuntimeComment(expr build.Expr) bool {
+	for _, comment := range expr.Comment().Suffix {
+		if strings.Contains(strings.ToLower(comment.Token), "runtime") {
+			return true
+		}
+	}
+	return false
+}
+
+// printCommands prints, for each key in the deps map, a buildozer command
+// to remove that entry from the deps attribute of the rule identified by label.
+// Returns true if at least one command was printed, or false otherwise.
+func printCommands(label string, deps map[string]bool) (anyCommandPrinted bool) {
+	buildFileName, pkg, ruleName := edit.InterpretLabel(label)
+	depsExpr := getDepsExpr(buildFileName, ruleName)
+	for _, li := range edit.AllLists(depsExpr) {
+		for _, elem := range li.List {
+			for dep := range deps {
+				str, ok := elem.(*build.StringExpr)
+				if ok && edit.ShortenLabel(str.Value, pkg) == edit.ShortenLabel(dep, pkg) {
```

Comment:
Maybe instead if `==` it make sense to use`strings.Contains(edit.ShortenLabel(dep, pkg), edit.ShortenLabel(str.Value, pkg))`? Sometimes, in BUILD file people use shorter versions of labels. For example, I recently faced a case, where `str.Value = org_apache_commons_commons_lang3//jar` and `dep = @@org_apache_commons_commons_lang3//jar:jar`. 

Or maybe it make sense instead of `.ShortenLabel` use some modified version of `.LabelsEqual`?